bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java

```diff
@@ -0,0 +1,979 @@
+    }
+
+    /**
+     * HTTP GET a Resources object, for a given resource Reference, from the Hue Bridge. The reference is a class
+     * comprising a resource type and an id. If the id is a specific resource id then only the one specific resource
+     * is returned, whereas if it is null then all resources of the given resource type are returned.
+     *
+     * It wraps the getResourcesImpl() method in a try/catch block, and transposes any HttpUnAuthorizedException into an
+     * ApiException. Such transposition should never be required in reality since by the time this method is called, the
+     * connection will surely already have been authorised.
+     *
+     * @param reference the Reference class to get.
+     * @return a Resource object containing either a list of Resources or a list of Errors.
+     * @throws ApiException if anything fails.
+     */
+    public Resources getResources(ResourceReference reference) throws ApiException {
+        sleepDuringRestart();
+        if (onlineState == State.CLOSED) {
+            throw new ApiException("getResources() offline");
+        }
+        try {
+            return getResourcesImpl(reference);
+        } catch (HttpUnauthorizedException e) {
+            throw new ApiException("getResources() unauthorized error", e);
+        }
+    }
+
+    /**
+     * Internal method to send an HTTP 2 GET request to the Hue Bridge and process its response.
+     *
+     * @param reference the Reference class to get.
+     * @return a Resource object containing either a list of Resources or a list of Errors.
+     * @throws ApiException if the communication failed, or an unexpected result occurred.
+     * @throws HttpUnauthorizedException if the request was refused as not authorised or forbidden.
+     */
+    private Resources getResourcesImpl(ResourceReference reference) throws ApiException, HttpUnauthorizedException {
+        Session session = http2Session;
+        if (Objects.isNull(session) || session.isClosed()) {
+            throw new ApiException("HTTP 2 session is null or closed");
+        }
+        throttle();
+        String url = getUrl(reference);
+        HttpFields fields = new HttpFields();
+        fields.put(HttpHeader.ACCEPT, MediaType.APPLICATION_JSON);
+        fields.put(HttpHeader.USER_AGENT, APPLICATION_ID);
+        fields.put(APPLICATION_KEY, applicationKey);
+        HeadersFrame headers = new HeadersFrame(
+                new MetaData.Request("GET", new HttpURI(url), HttpVersion.HTTP_2, fields), null, true);
+        ContentAdapter adapter = new ContentAdapter(this);
+        Completable<@Nullable Stream> completable = new Completable<>();
+        logger.trace("GET {} HTTP/2", url);
+        try {
+            session.newStream(headers, completable, adapter);
+            Objects.requireNonNull(completable.get(TIMEOUT_SECONDS, TimeUnit.SECONDS));
+            String content = adapter.completable.get(TIMEOUT_SECONDS, TimeUnit.SECONDS).trim();
```

Comment:
In the meantime I found out the suggested API was added in versions of Jetty newer than what we're using. I still think there's improvement potential, though, particularly in variable naming and code duplication. My suggestion would be something like this (compile time tested only, includes the diff above):

```diff
diff --git a/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java b/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java
index cd6189e68b..958ca4bc2c 100644
--- a/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java
+++ b/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java
@@ -20,6 +20,7 @@ import java.nio.charset.StandardCharsets;
 import java.time.Duration;
 import java.time.Instant;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
@@ -29,6 +30,7 @@ import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
 
 import javax.ws.rs.core.MediaType;
 
@@ -56,7 +58,6 @@ import org.eclipse.jetty.http2.frames.PingFrame;
 import org.eclipse.jetty.http2.frames.ResetFrame;
 import org.eclipse.jetty.util.Callback;
 import org.eclipse.jetty.util.Promise.Completable;
-import org.eclipse.jetty.util.ssl.SslContextFactory;
 import org.openhab.binding.hue.internal.dto.CreateUserRequest;
 import org.openhab.binding.hue.internal.dto.SuccessResponse;
 import org.openhab.binding.hue.internal.dto.clip2.BridgeConfig;
@@ -94,29 +95,19 @@ import com.google.gson.JsonParser;
 public class Clip2Bridge implements Closeable {
 
     /**
-     * Interface for processing adapter errors. It handles fatal errors by implementing the fatalError() method.
+     * Enum of potential fatal HTTP 2 session/stream errors.
      *
      * @author Andrew Fiddian-Green - Initial Contribution
      */
-    private static interface AdapterErrorHandler {
-
-        /**
-         * Enum of potential fatal HTTP 2 session/stream errors.
-         *
-         * @author Andrew Fiddian-Green - Initial Contribution
-         */
-        public static enum Error {
-            CLOSED,
-            ERROR,
-            FAILURE,
-            TIMEOUT,
-            RESET,
-            IDLE,
-            GO_AWAY,
-            UNAUTHORIZED;
-        }
-
-        public void fatalError(Error error);
+    private static enum HttpError {
+        CLOSED,
+        ERROR,
+        FAILURE,
+        TIMEOUT,
+        RESET,
+        IDLE,
+        GO_AWAY,
+        UNAUTHORIZED;
     }
 
     /**
@@ -131,28 +122,24 @@ public class Clip2Bridge implements Closeable {
      *
      * @author Andrew Fiddian-Green - Initial Contribution
      */
-    private abstract static class BaseAdapter extends Stream.Listener.Adapter implements AdapterErrorHandler {
-
-        protected final Clip2Bridge owner;
-        protected final List<String> strings = new ArrayList<>();
-        protected final CompletableFuture<String> completable = new CompletableFuture<>();
+    private abstract class BaseAdapter<T> extends Stream.Listener.Adapter {
+        protected final CompletableFuture<T> completable = new CompletableFuture<T>();
 
         private String contentType = "UNDEFINED";
 
-        protected BaseAdapter(Clip2Bridge owner) {
-            this.owner = owner;
-        }
-
-        @Override
-        public void fatalError(Error error) {
+        protected void fatalError(HttpError error) {
             Exception e;
-            if (Error.UNAUTHORIZED.equals(error)) {
+            if (HttpError.UNAUTHORIZED.equals(error)) {
                 e = new HttpUnauthorizedException("HTTP 2 request not authorized");
             } else {
                 e = new ApiException("HTTP 2 stream " + error.toString().toLowerCase());
             }
             completable.completeExceptionally(e);
-            owner.fatalErrorDelayed(this, error);
+            fatalErrorDelayed(this, error);
+        }
+
+        public T awaitResult() throws ExecutionException, InterruptedException, TimeoutException {
+            return completable.get(TIMEOUT_SECONDS, TimeUnit.SECONDS);
         }
 
         /**
@@ -177,7 +164,7 @@ public class Clip2Bridge implements Closeable {
                 switch (httpStatus) {
                     case HttpStatus.UNAUTHORIZED_401:
                     case HttpStatus.FORBIDDEN_403:
-                        fatalError(Error.UNAUTHORIZED);
+                        fatalError(HttpError.UNAUTHORIZED);
                     default:
                 }
                 contentType = responseMetaData.getFields().get(HttpHeader.CONTENT_TYPE).toLowerCase();
@@ -198,20 +185,16 @@ public class Clip2Bridge implements Closeable {
      *
      * @author Andrew Fiddian-Green - Initial Contribution
      */
-    private static class ContentAdapter extends BaseAdapter {
-
-        protected ContentAdapter(Clip2Bridge owner) {
-            super(owner);
-        }
+    private class ContentAdapter extends BaseAdapter<String> {
 
         @Override
         public void onData(@Nullable Stream stream, @Nullable DataFrame frame, @Nullable Callback callback) {
             Objects.requireNonNull(frame);
             Objects.requireNonNull(callback);
             synchronized (this) {
-                strings.add(StandardCharsets.UTF_8.decode(frame.getData()).toString());
+                String content = StandardCharsets.UTF_8.decode(frame.getData()).toString();
                 if (frame.isEndStream() && !completable.isDone()) {
-                    completable.complete(String.join("", strings));
+                    completable.complete(content.trim());
                 }
             }
             callback.succeeded();
@@ -219,13 +202,13 @@ public class Clip2Bridge implements Closeable {
 
         @Override
         public boolean onIdleTimeout(@Nullable Stream stream, @Nullable Throwable x) {
-            fatalError(Error.IDLE);
+            fatalError(HttpError.IDLE);
             return true;
         }
 
         @Override
         public void onTimeout(@Nullable Stream stream, @Nullable Throwable x) {
-            fatalError(Error.TIMEOUT);
+            fatalError(HttpError.TIMEOUT);
         }
     }
 
@@ -247,15 +230,11 @@ public class Clip2Bridge implements Closeable {
      *
      * @author Andrew Fiddian-Green - Initial Contribution
      */
-    private static class EventAdapter extends BaseAdapter {
-
-        protected EventAdapter(Clip2Bridge owner) {
-            super(owner);
-        }
+    private class EventAdapter extends BaseAdapter<Boolean> {
 
         @Override
         public void onClosed(@Nullable Stream stream) {
-            fatalError(Error.CLOSED);
+            fatalError(HttpError.CLOSED);
         }
 
         @Override
@@ -263,17 +242,14 @@ public class Clip2Bridge implements Closeable {
             Objects.requireNonNull(frame);
             Objects.requireNonNull(callback);
             if (!completable.isDone()) {
-                completable.complete(Boolean.toString(true));
+                completable.complete(Boolean.TRUE);
             }
             synchronized (this) {
-                for (String line : StandardCharsets.UTF_8.decode(frame.getData()).toString().split("\\R")) {
-                    if (line.startsWith("data: ")) {
-                        strings.add(line.substring(6));
-                    }
-                }
-                if (!strings.isEmpty()) {
-                    owner.onEventData(String.join("", strings));
-                    strings.clear();
+                String event = Arrays.stream(StandardCharsets.UTF_8.decode(frame.getData()).toString().split("\\R"))
+                        .filter(line -> line.startsWith("data: ")).map(line -> line.substring(6))
+                        .collect(Collectors.joining(""));
+                if (!event.isEmpty()) {
+                    onEventData(event);
                 }
             }
             callback.succeeded();
@@ -286,7 +262,7 @@ public class Clip2Bridge implements Closeable {
 
         @Override
         public void onReset(@Nullable Stream stream, @Nullable ResetFrame frame) {
-            fatalError(Error.RESET);
+            fatalError(HttpError.RESET);
         }
     }
 
@@ -303,32 +279,21 @@ public class Clip2Bridge implements Closeable {
      *
      * @author Andrew Fiddian-Green - Initial Contribution
      */
-    private static class SessionAdapter extends Session.Listener.Adapter implements AdapterErrorHandler {
-
-        private final Clip2Bridge owner;
-
-        protected SessionAdapter(Clip2Bridge owner) {
-            this.owner = owner;
-        }
-
-        @Override
-        public void fatalError(Error error) {
-            owner.fatalErrorDelayed(this, error);
-        }
+    private class SessionAdapter extends Session.Listener.Adapter {
 
         @Override
         public void onClose(@Nullable Session session, @Nullable GoAwayFrame frame) {
-            fatalError(Error.CLOSED);
+            fatalErrorDelayed(this, HttpError.CLOSED);
         }
 
         @Override
         public void onFailure(@Nullable Session session, @Nullable Throwable failure) {
-            fatalError(Error.FAILURE);
+            fatalErrorDelayed(this, HttpError.FAILURE);
         }
 
         @Override
         public void onGoAway(@Nullable Session session, @Nullable GoAwayFrame frame) {
-            fatalError(Error.GO_AWAY);
+            fatalErrorDelayed(this, HttpError.GO_AWAY);
         }
 
         @Override
@@ -338,7 +303,7 @@ public class Clip2Bridge implements Closeable {
 
         @Override
         public void onPing(@Nullable Session session, @Nullable PingFrame frame) {
-            owner.checkAliveOk();
+            checkAliveOk();
             if (Objects.nonNull(session) && Objects.nonNull(frame) && !frame.isReply()) {
                 session.ping(new PingFrame(true), Callback.NOOP);
             }
@@ -346,7 +311,7 @@ public class Clip2Bridge implements Closeable {
 
         @Override
         public void onReset(@Nullable Session session, @Nullable ResetFrame frame) {
-            fatalError(Error.RESET);
+            fatalErrorDelayed(this, HttpError.RESET);
         }
     }
 
@@ -459,7 +424,7 @@ public class Clip2Bridge implements Closeable {
             session.ping(new PingFrame(false), Callback.NOOP);
         }
         if (Instant.now().isAfter(sessionExpireTime)) {
-            fatalError(this, AdapterErrorHandler.Error.TIMEOUT);
+            fatalError(this, HttpError.TIMEOUT);
         }
     }
 
@@ -526,14 +491,14 @@ public class Clip2Bridge implements Closeable {
      * @param cause the entity that caused this method to be called.
      * @param error the type of error.
      */
-    private void fatalError(Object cause, BaseAdapter.Error error) {
+    private void fatalError(Object cause, HttpError error) {
         if (restarting || (onlineState == State.CLOSED)) {
             return;
         }
         String causeId = cause.getClass().getSimpleName();
         if (cause instanceof ContentAdapter) {
             logger.debug("fatalError() {} {} ignoring", causeId, error);
-        } else if (error == AdapterErrorHandler.Error.GO_AWAY) {
+        } else if (error == HttpError.GO_AWAY) {
             logger.debug("fatalError() {} {} reconnecting", causeId, error);
 
             // close
@@ -568,7 +533,7 @@ public class Clip2Bridge implements Closeable {
      * @param cause the entity that caused this method to be called.
      * @param error the type of error.
      */
-    private void fatalErrorDelayed(Object cause, BaseAdapter.Error error) {
+    private void fatalErrorDelayed(Object cause, HttpError error) {
         bridgeHandler.getScheduler().schedule(() -> fatalError(cause, error), 1, TimeUnit.SECONDS);
     }
 
@@ -597,6 +562,26 @@ public class Clip2Bridge implements Closeable {
         }
     }
 
+    private HeadersFrame prepareHeaders(String url, String acceptMediaType) {
+        return prepareHeaders(url, acceptMediaType, "GET", -1, null);
+    }
+
+    private HeadersFrame prepareHeaders(String url, String acceptMediaType, String method, long contentLength,
+            @Nullable String contentType) {
+        HttpFields fields = new HttpFields();
+        fields.put(HttpHeader.ACCEPT, acceptMediaType);
+        if (contentType != null) {
+            fields.put(HttpHeader.CONTENT_TYPE, contentType);
+        }
+        if (contentLength >= 0) {
+            fields.putLongField(HttpHeader.CONTENT_LENGTH, contentLength);
+        }
+        fields.put(HttpHeader.USER_AGENT, APPLICATION_ID);
+        fields.put(APPLICATION_KEY, applicationKey);
+        return new HeadersFrame(new MetaData.Request(method, new HttpURI(url), HttpVersion.HTTP_2, fields), null,
+                contentLength <= 0);
+    }
+
     /**
      * Internal method to send an HTTP 2 GET request to the Hue Bridge and process its response.
      *
@@ -612,39 +597,34 @@ public class Clip2Bridge implements Closeable {
         }
         throttle();
         String url = getUrl(reference);
-        HttpFields fields = new HttpFields();
-        fields.put(HttpHeader.ACCEPT, MediaType.APPLICATION_JSON);
-        fields.put(HttpHeader.USER_AGENT, APPLICATION_ID);
-        fields.put(APPLICATION_KEY, applicationKey);
-        HeadersFrame headers = new HeadersFrame(
-                new MetaData.Request("GET", new HttpURI(url), HttpVersion.HTTP_2, fields), null, true);
-        ContentAdapter adapter = new ContentAdapter(this);
-        Completable<@Nullable Stream> completable = new Completable<>();
+        HeadersFrame headers = prepareHeaders(url, MediaType.APPLICATION_JSON);
+        ContentAdapter adapter = new ContentAdapter();
         logger.trace("GET {} HTTP/2", url);
         try {
-            session.newStream(headers, completable, adapter);
-            Objects.requireNonNull(completable.get(TIMEOUT_SECONDS, TimeUnit.SECONDS));
-            String content = adapter.completable.get(TIMEOUT_SECONDS, TimeUnit.SECONDS).trim();
+            session.newStream(headers, null, adapter);
+            String content = adapter.awaitResult();
             String contentType = adapter.getContentType();
             logger.trace("HTTP/2 200 OK (Content-Type: {}) << {}", contentType, content);
             if (!MediaType.APPLICATION_JSON.equals(contentType)) {
                 throw new ApiException("Unexpected Content-Type: " + contentType);
             }
-            Resources resources = Objects.requireNonNull(jsonParser.fromJson(content, Resources.class));
-            if (logger.isDebugEnabled()) {
-                resources.getErrors().forEach(error -> logger.debug("Resources error:{}", error));
+            try {
+                Resources resources = Objects.requireNonNull(jsonParser.fromJson(content, Resources.class));
+                if (logger.isDebugEnabled()) {
+                    resources.getErrors().forEach(error -> logger.debug("Resources error:{}", error));
+                }
+                return resources;
+            } catch (JsonParseException e) {
+                throw new ApiException("Parsing error", e);
             }
-            return resources;
         } catch (ExecutionException e) {
-            Throwable e2 = e.getCause();
-            if (e2 instanceof HttpUnauthorizedException) {
-                throw (HttpUnauthorizedException) e2;
+            Throwable cause = e.getCause();
+            if (cause instanceof HttpUnauthorizedException) {
+                throw (HttpUnauthorizedException) cause;
             }
             throw new ApiException("Error sending request", e);
         } catch (InterruptedException | TimeoutException e) {
             throw new ApiException("Error sending request", e);
-        } catch (JsonParseException e) {
-            throw new ApiException("Parsing error", e);
         }
     }
 
@@ -755,22 +735,17 @@ public class Clip2Bridge implements Closeable {
             return;
         }
         logger.debug("openEventStream()");
-        HttpFields fields = new HttpFields();
-        fields.put(HttpHeader.ACCEPT, MediaType.SERVER_SENT_EVENTS);
-        fields.put(HttpHeader.USER_AGENT, APPLICATION_ID);
-        fields.put(APPLICATION_KEY, applicationKey);
-        MetaData.Request request = new MetaData.Request("GET", new HttpURI(eventUrl), HttpVersion.HTTP_2, fields);
-        HeadersFrame headers = new HeadersFrame(request, null, true);
-        EventAdapter adapter = new EventAdapter(this);
-        Completable<@Nullable Stream> completable = new Completable<>();
+        HeadersFrame headers = prepareHeaders(eventUrl, MediaType.SERVER_SENT_EVENTS);
+        EventAdapter adapter = new EventAdapter();
         logger.trace("GET {} HTTP/2", eventUrl);
         Stream stream = null;
         try {
-            session.newStream(headers, completable, adapter);
-            stream = Objects.requireNonNull(completable.get(TIMEOUT_SECONDS, TimeUnit.SECONDS));
+            Completable<@Nullable Stream> streamPromise = new Completable<>();
+            session.newStream(headers, streamPromise, adapter);
+            stream = Objects.requireNonNull(streamPromise.get(TIMEOUT_SECONDS, TimeUnit.SECONDS));
             stream.setIdleTimeout(0);
             stream.setAttribute(EVENT_STREAM_ID, session);
-            adapter.completable.get(TIMEOUT_SECONDS, TimeUnit.SECONDS);
+            adapter.awaitResult();
         } catch (ExecutionException | InterruptedException | TimeoutException e) {
             if (Objects.nonNull(stream)) {
                 stream.reset(new ResetFrame(stream.getId(), 0), Callback.NOOP);
@@ -812,11 +787,11 @@ public class Clip2Bridge implements Closeable {
         }
         logger.debug("openSession()");
         InetSocketAddress address = new InetSocketAddress(hostName, 443);
-        SessionAdapter adapter = new SessionAdapter(this);
-        Completable<@Nullable Session> completable = new Completable<>();
+        SessionAdapter adapter = new SessionAdapter();
         try {
-            http2Client.connect(http2Client.getBean(SslContextFactory.class), address, adapter, completable);
-            http2Session = Objects.requireNonNull(completable.get(TIMEOUT_SECONDS, TimeUnit.SECONDS));
+            Completable<@Nullable Session> sessionPromise = new Completable<>();
+            http2Client.connect(address, adapter, sessionPromise);
+            http2Session = Objects.requireNonNull(sessionPromise.get(TIMEOUT_SECONDS, TimeUnit.SECONDS));
             checkAliveOk(); // initialise the session timeout window
         } catch (InterruptedException | ExecutionException | TimeoutException e) {
             throw new ApiException("Error opening HTTP 2 session", e);
@@ -842,37 +817,34 @@ public class Clip2Bridge implements Closeable {
         String json = jsonParser.toJson(resource);
         ByteBuffer jsonBytes = ByteBuffer.wrap(json.getBytes(StandardCharsets.UTF_8));
         String url = getUrl(new ResourceReference().setId(resource.getId()).setType(resource.getType()));
-        HttpFields fields = new HttpFields();
-        fields.put(HttpHeader.ACCEPT, MediaType.APPLICATION_JSON);
-        fields.put(HttpHeader.USER_AGENT, APPLICATION_ID);
-        fields.put(HttpHeader.CONTENT_TYPE, MediaType.APPLICATION_JSON);
-        fields.putLongField(HttpHeader.CONTENT_LENGTH, jsonBytes.capacity());
-        fields.put(APPLICATION_KEY, applicationKey);
-        HeadersFrame headers = new HeadersFrame(
-                new MetaData.Request("PUT", new HttpURI(url), HttpVersion.HTTP_2, fields), null, false);
-        ContentAdapter adapter = new ContentAdapter(this);
-        Completable<@Nullable Stream> completable = new Completable<>();
+        HeadersFrame headers = prepareHeaders(url, MediaType.APPLICATION_JSON, "PUT", jsonBytes.capacity(),
+                MediaType.APPLICATION_JSON);
+        ContentAdapter adapter = new ContentAdapter();
         logger.trace("PUT {} HTTP/2 >> {}", url, json);
-        String content = "";
         try {
-            session.newStream(headers, completable, adapter);
-            Stream stream = Objects.requireNonNull(completable.get(TIMEOUT_SECONDS, TimeUnit.SECONDS));
+            Completable<@Nullable Stream> streamPromise = new Completable<>();
+            session.newStream(headers, streamPromise, adapter);
+
+            Stream stream = Objects.requireNonNull(streamPromise.get(TIMEOUT_SECONDS, TimeUnit.SECONDS));
             stream.data(new DataFrame(stream.getId(), jsonBytes, true), Callback.NOOP);
-            content = adapter.completable.get(TIMEOUT_SECONDS, TimeUnit.SECONDS).trim();
+
+            String content = adapter.awaitResult();
             String contentType = adapter.getContentType();
             logger.trace("HTTP/2 200 OK (Content-Type: {}) << {}", contentType, content);
             if (!MediaType.APPLICATION_JSON.equals(contentType)) {
                 throw new ApiException("Unexpected Content-Type: " + contentType);
             }
-            Resources resources = Objects.requireNonNull(jsonParser.fromJson(content, Resources.class));
-            if (logger.isDebugEnabled()) {
-                resources.getErrors().forEach(error -> logger.debug("Resources error:{}", error));
+            try {
+                Resources resources = Objects.requireNonNull(jsonParser.fromJson(content, Resources.class));
+                if (logger.isDebugEnabled()) {
+                    resources.getErrors().forEach(error -> logger.debug("Resources error:{}", error));
+                }
+            } catch (JsonParseException e) {
+                logger.warn("putResourceHttp2Impl() error parsing JSON response:{}", content);
+                throw new ApiException("Parsing error", e);
             }
         } catch (ExecutionException | InterruptedException | TimeoutException e) {
             throw new ApiException("Error sending request", e);
-        } catch (JsonParseException e) {
-            logger.warn("putResourceHttp2Impl() error parsing JSON response:{}", content);
-            throw new ApiException("Parsing error", e);
         }
     }
 
@@ -902,13 +874,13 @@ public class Clip2Bridge implements Closeable {
             throw new ApiException("HTTP processing error", e);
         }
         int httpStatus = contentResponse.getStatus();
-        json = contentResponse.getContentAsString().trim();
-        logger.trace("HTTP/1.1 {} {} << {}", httpStatus, contentResponse.getReason(), json);
+        String responseJson = contentResponse.getContentAsString().trim();
+        logger.trace("HTTP/1.1 {} {} << {}", httpStatus, contentResponse.getReason(), responseJson);
         if (httpStatus != HttpStatus.OK_200) {
             throw new ApiException("HTTP bad response");
         }
         try {
-            List<SuccessResponse> entries = jsonParser.fromJson(json, SuccessResponse.GSON_TYPE);
+            List<SuccessResponse> entries = jsonParser.fromJson(responseJson, SuccessResponse.GSON_TYPE);
             if (Objects.nonNull(entries) && !entries.isEmpty()) {
                 SuccessResponse response = entries.get(0);
                 Map<String, Object> responseSuccess = response.success;
```

Reply:
> The interaction between completable and adapter.completable is somewhat confusing.
> Is it really necessary to have 2 separate completables here?

I think yes.
- The first completable checks if the HTTP stream was created; and it fails due to `Objects.requireNonNull()`; probably a network error.
- The second completable checks for an actual response over the stream; and it fails due to request time out; probably a server error.

Reply:
> In the meantime I found out the suggested API was added in versions of Jetty newer than what we're using.

Indeed. Jetty has a breaking change in its HTTP/2 implementation between Jetty v9 and v10. So when the OH Core is (finally) updated to Jetty v11 it will require a rewrite of this class.

Reply:
A request timeout (no data arrives) likely means 'network error' as well. The point is: as long as we're not doing anything with the stream, I don't see why we'd need to await it's creation. Either it times out (then the data promise will time out as well), or it completes (then the data promise may or may not time out). The only use case I can see is failure in stream creation.
So maybe it makes sense to keep awaiting the stream promise; the potential confusion can be fought by variable naming (`streamPromise` instead of `completable` in my diff above) and encapsulation (the `awaitResult` method in my diff above).

Reply:
> -  http2Client.connect(http2Client.getBean(SslContextFactory.class), address, adapter, sessionPromise);
> + http2Client.connect(address, adapter, sessionPromise);

Are you 100% sure that we don't need the SslContextFactory.class here? The Hue Bridge uses HTTPS SSL streams only.

Reply:
Not 100% sure, but reasonably certain that it will have a suitable default one, especially since the one you're passing in is taken from the client anyway.

Reply:
@maniac103 I found it incredibly difficult to track so many change request in one request. In future could you please split into multiple requests? Nevertheless, I think that I (will) have taken over most of your requests. But NOT all of them!!

Reply:
Do you mean the diff? I was just hacking around a bit yesterday to make a bit less theoretical suggestions and see what's actually possible to do ;-) Or did you mean something else?
In any case, I'll have another look once you push new code.